tsinfer/inference.py

```diff
@@ -890,23 +980,77 @@ def __init__(
             [inference_site_position, [sample_data.sequence_length]]
         )
         self.position_map[0] = 0
+        self.recombination = np.zeros(self.num_sites)  # TODO: reduce len by 1
+        self.mismatch = np.zeros(self.num_sites)
+
+        # Set the canonical 'recombination' and 'mismatch' arrays to appropriate values
+        # NB 'recombination' must be of length num_sites-1, mismatch of length num_sites
+
+        # First set defaults if None provided
+        if recombination_rate is None and recombination is None:
+            if mismatch_ratio is None and mismatch is None:
+                # Special case: revert to tsinfer 0.1 behaviour with no mismatch allowed
+                mismatch = np.full(self.num_sites, 1e-20)  # A miniscule value
+                # arbitrary, large relative to mismatch, but < 1
+                if self.num_sites == 0:
+                    recombination = []
+                else:
+                    recombination = np.full(self.num_sites - 1, 1e-2)
+                # Shouldn't need insane precision, although 3 is not enough in trials
+                if precision is None:
+                    precision = 5
+            else:
+                raise ValueError("Cannot use mismatch without setting recombination")
+        if mismatch_ratio is None and mismatch is None:
+            # Default if recombination but no mismatch params have been specified
+            mismatch_ratio = 1.0
+
+        assert not (recombination_rate is None and recombination is None)
+        assert not (mismatch_ratio is None and mismatch is None)
+
+        # Set recombination from recombination_rate
+        if recombination is None:
+            try:
+                recombination = np.diff(
```

Comment:
Is this to support both pre and post 1.0 msprime RateMaps? If so, we should pull this out into a function so we can test it rather than being deeply nested in here.

Reply:
No, it's so that we can support either a floating point value *or* a RateMap (or indeed, anything that provides a `.get_cumulative_mass()` method). If no such method exists, we treat recombination_rate as a float.

Reply:
OK, let's make a ``parse_recombination_rate`` function then to abstract this out.

Reply:
Done, although I called it `recombination_rate_to_probs`